### ISSUE TITLE
engine fails to start

### ISSUE THREAD

**Issue by dodler** [sqlalchemy/sqlalchemy#6083]
Hi, I got this unexpected behaviour

I got sqlalchemy.exc.ArgumentError: Can't load plugin: sqlalchemy.dialects:driver
 when creating connection to postgres. 

I'm using this this address to connect postgres://postgres:1123@database:5432/postgres, 
but after update to 1.4.1 it started to produce the above mentioned error. 

Update to 1.3.0 helped me to resolve this error

**Comment by rabidaudio** [comment]
We had a similar issue start today. For some reason, changing the database url from `postgres://` to `postgresql://` solved the problem.

**Comment by skelliest** [comment]
Also getting this in our builds as of Tue, 16 Mar 2021 21:01:00 GMT

**Comment by jvanasco** [comment]
The "postgres" dialect was deprecated from SQLAlchemy in version 0.6, which was available for testing purposes against the public repository since August 2009, and then rolled out to a public release in February 2010.  For the past 11-12 years, using the connection string with "postgres" and not "postgresql" has generated a warning the dialect was deprecated, and developers should migrate to the correct dialect, "postgresql".

Legacy projects have had an extremely generous timeframe of over ten years across eight release cycles to migrate from the deprecated term "postgres" to "postgresql".  Projects started since SQLAlchemy v0.6 should never have used this dialect name to begin with.

Unfortunately, many people have ignored the warnings that SQLAlchemy has generated and are now considering this as an issue.  This is not unexpected. This is not an issue. 

* Aug 6, 2009 - Initial Deprecation  on v0.6 branch https://github.com/sqlalchemy/sqlalchemy/blob/8fc5005dfe3eb66a46470ad8a8c7b95fc4d6bdca/lib/sqlalchemy/dialects/postgres.py

* January 2016 - Remove the long deprecated "postgres" module, issue warnings and translate within dialects/__init__.py: https://github.com/sqlalchemy/sqlalchemy/commit/c59bf0007f0148b73c80a3d6b86d6c66ae0ab422#diff-a453b1c3c392ab7bd09e97ab080042e9e4d1485f107b99011a39729c561004ba

* PyPi release history - v0.6 April 2010, betas in February 2010 https://pypi.org/project/SQLAlchemy/#history

**Comment by zzzeek** [comment]
can we "dupe" all these ?  there were other issues opened w/ people surprised by the name removal right?

**Comment by jvanasco** [comment]
@zzzeek I think this is the first one against SQLAlchemy itself. The other one I know of is against flask-sqlalchemy, and they made a comment on @CaselIT's commit.

See:
* https://github.com/pallets/flask-sqlalchemy/issues/929
* https://github.com/sqlalchemy/sqlalchemy/commit/a9b068ae564e5e775e312373088545b75aeaa1b0#r48347973

**Comment by zzzeek** [comment]
> Also getting this in our builds as of Tue, 16 Mar 2021 21:01:00 GMT. Can confirm updating url will fix, but is not viable for us as url is stored as a secret

you would need to implement a translator for your URLs to repair this, at the point at which your URL comes from some kind of secret and is passed to create_engine() without encryption.  

you can also modify URLs using the [CreateEnginePlugin](https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.CreateEnginePlugin).   if you write a plugin, an alternative to setting it up as an entrypoint is registering it with `sqlalchemy.dialects.plugins.register("mymodule", "plugin_name", "PluginClassName")`.

**Comment by rabidaudio** [comment]
To be fair, many environments (including Heroku last I was aware), supply database URLs using the `postgres://` schema, so this is not necessarily a decision in developer control.

This is admittedly anecdotal, but I see `postgres://` much more commonly used in the industry than `postgresql://`

**Comment by skelliest** [comment]
Yeah, that's fine. We've been using SQLAlchemy since 2017 and never received a deprecation warning about the URL, but 🤷  we can work around.

**Comment by jvanasco** [comment]
> To be fair, many environments (including Heroku last I was aware), supply database URLs using the `postgres://` schema, so this is not necessarily a decision in developer control.

I haven't used Heroku in years. When I ran into issues such as this on Heroku or other PAAS systems, I would contact our support specialist or their developer evangelists and ask them to correct the problem.

**Comment by zzzeek** [comment]
> To be fair, many environments (including Heroku last I was aware), supply database URLs using the `postgres://` schema, so this is not necessarily a decision in developer control.
> 
> This is admittedly anecdotal, but I see `postgres://` much more commonly used in the industry than `postgresql://`

this name is probably the oldest deprecated thing in all of SQLAlchemy.   There's been no controversy or discussion about removing the name for 6, 7, 8 years, emitting a warning for almost the whole time, until we removed it.    I get hit by projects removing names all the time, where they've deprecated something and I wasn't checking (lots of things in pytest, sphinx, etc) , but that's on me.        In this case Heroku needs to change this and they should have changed it years ago when they see the warning in their logs.

**Comment by jvanasco** [comment]
> We've been using SQLAlchemy since 2017 and never received a deprecation warning about the URL

Starting in Python 2.7/3.2, DeprecationWarnings have been invisible by default.  See https://docs.python.org/3/whatsnew/2.7.html#changes-to-the-handling-of-deprecation-warnings

Under Python3, `unittest` automatically re-enables them. I think one of the latter Python2.7 versions re-enables them under unittest as well, but I don't think earlier versions did.  Most developers periodically run test suites and CI with deprecation warnings enabled (`python -Wdefault`) to catch these concerns.

**Comment by zzzeek** [comment]
im really not understanding how deprecation warning works

```
$ python
Python 3.8.3 (default, May 23 2020, 16:34:37) 
[GCC 9.3.1 20200408 (Red Hat 9.3.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sqlalchemy import util
>>> util.warn_deprecated("asdfsd", "1.4")
>>> # ok, no warning, so it's silent.  but then....
>>> from sqlalchemy import select
>>> select().c
<stdin>:1: SADeprecationWarning: The SelectBase.c and SelectBase.columns attributes are deprecated and will be removed in a future release; these attributes implicitly create a subquery that should be explicit.  Please call SelectBase.subquery() first in order to create a subquery, which then contains this attribute.  To access the columns that this SELECT object SELECTs from, use the SelectBase.selected_columns attribute. (deprecated since: 1.4)
<sqlalchemy.sql.base.ImmutableColumnCollection object at 0x7fd0a13de9f0>
>>> # WAT

```

**Comment by jvanasco** [comment]
> im really not understanding how deprecation warning works

No one really does anymore.

IIRC, the 2.7/3.2 switch - and latter updates - were trying to get this behavior:

* Don't show any deprecations when you run a program, because end-users keep thinking everything will break.
* Do show deprecations in interactive interpreters
* Maybe show deprecations in __main__, but I'm not sure if it's for an app or an interactive only session
* Show deprecations when running tests
* A bunch of other rules.

9/10 times I think I'm supposed to see a warning, I don't.  So... `python -Wdefault`

tldr; I f****** hated this change, and still do.

**Comment by jvanasco** [comment]
@zzzeek PEP-565 goes deep into this logic and behavior. 
 I knew i had read it somewhere!  https://www.python.org/dev/peps/pep-0565/

**Comment by dodler** [comment]
Got it, thanks a lot.

**Comment by CaselIT** [comment]
I wonder if we should use UserWarning instead of DeprecationWarning as base class for warnings...
Does it mean that people are also not seeing the other warning that we have (and 1.4 has quite a lot of them)?


> This is admittedly anecdotal, but I see `postgres://` much more commonly used in the industry than `postgresql://`

I have the opposite experience. The [postgresql docuementation](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING) uses only `postgresql://`, searching for `postgresql conection uri` seem to show mostly `postgresql://`. I have also used aws and azure managed instances and both use `postgresql://` as connection string.

**Comment by mble** [comment]
Hi folks – I work on the Heroku team that manages our data services, including Heroku Postgres. Given that `libpq` itself accepts _either_ `postgres` or `postgresql` as the URI scheme (https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING), removing this entirely feels like a misstep until `libpq` removes support for that scheme.

I personally take `libpq` as the reference implementation for Postgres clients.

> In this case Heroku needs to change this and they should have changed it years ago when they see the warning in their logs

We don't perform large scale log analysis of customer logs out of privacy considerations.

However, all that said, I'll look into updating the scheme to `postgresql://` in a sustainable way. Can't provide timelines etc.

**Comment by zzzeek** [comment]
> Hi folks – I work on the Heroku team that manages our data services, including Heroku Postgres. Given that `libpq` itself accepts _either_ `postgres` or `postgresql` as the URI scheme (https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING), removing this entirely feels like a misstep until `libpq` removes support for that scheme.

noted, but overall SQLAlchemy URLs have only one way to specify a particular database backend.  there's no URL scheme that accepts two different terms.  libpq is an implementation detail.

> 
> I personally take `libpq` as the reference implementation for Postgres clients.

We've had 6-7 years or so for anyone to have any comment about this change and there weren't any, so we went ahead with finally removing it.


> 
> We don't perform large scale log analysis of customer logs out of privacy considerations.

If this URL scheme is somehow hardcoded in Heroku's software that implies that Heroku itself is calling upon `create_engine()` somewhere, which means Herokus own test suites would have been seeing this warning all over the place.   No customer log analysis should be required.


> 
> However, all that said, I'll look into updating the scheme to `postgresql://` in a sustainable way. Can't provide timelines etc.

thanks!

**Comment by sigmavirus24** [comment]
@zzzeek I think there's a misunderstanding. Heroku Postgresql is a product offered and which uses the old scheme when providing URIs to customers of the service. This means anyone using SQLAlchemy as a Heroku customer is broken with 1.4.x

**Comment by jvanasco** [comment]
> Does it mean that people are also not seeing the other warning that we have (and 1.4 has quite a lot of them)?

AFAIK, The warnings will not show on Python27 unless you explicitly enable them on the app or test harness. On Py3, they must be explicitly enabled on apps but may not need to be enabled on the test harness. iirc: unittest, pytest and nose all enable warnings on py3.

Did I mention how much I really hate this change in Python core?  It really tripped me up about 10 years ago, and changed the way I check apps.

**Comment by zzzeek** [comment]
> @zzzeek I think there's a misunderstanding. Heroku Postgresql is a product offered and which uses the old scheme when providing URIs to customers of the service. This means anyone using SQLAlchemy as a Heroku customer is broken with 1.4.x

can you clarify "providing URIs" ?  are the customers writing code and this URI is a string, such that they can say uri = re.sub(r'^postgres', 'postgresql', uri) ?    if the customers are writing SQLAlchemy code, have they all been upgraded to SQLAlchemy 1.4 automatically or is this just customers who want to use SQLAlchemy 1.4 are upset they have to modify the URI string?   

the idea is, if im a customer, and I want to upgrade to SQLAlchemy 1.4 in the app I run on Heroku, then I also have to do this one workaround for now for the URI that Heroku gives me.  IMO that's not really breaking the world, it's only for customers who opt into using the new release.


> 
> Edit to add: I get the desire to remove something deprecated years ago. Also, pretty much every other language and Python ORM seems to handle both schemes. So maybe that helps understand the level of surprise felt.

**Comment by mble** [comment]
@zzzeek customers of Heroku Postgres are given a connection string to the resource they provision that uses the `postgres://` scheme. Customers can absolutely do some pre-processing to change the scheme. They are not being forced to use 1.4.x, only if they haven't pinned the version to <= 1.4.x.

Agreed it isn't breaking the world. More highlighting that due to the way the deprecation warnings were emitted many users may not have encountered them and known to implement the pre-processing step, where in virtually all other popular drivers this pre-processing step is not necessary.

**Comment by zzzeek** [comment]
the real bug here is Python's terrible inconsistency in warning display.  I right now don't understand the behavior in my comment at https://github.com/sqlalchemy/sqlalchemy/issues/6083#issuecomment-801514616 .

**Comment by cassiel257** [comment]
I was able to use this post to bypass the error on Heroku, at least:
```
import os
import re

uri = os.getenv("DATABASE_URL")  # or other relevant config var
if uri.startswith("postgres://"):
    uri = uri.replace("postgres://", "postgresql://", 1)
# rest of connection code using the connection string `uri`
```
https://help.heroku.com/ZKNTJQSK/why-is-sqlalchemy-1-4-x-not-connecting-to-heroku-postgres

**Comment by blacng** [comment]
> I was able to use this post to bypass the error on Heroku, at least:
> 
> ```
> import os
> import re
> 
> uri = os.getenv("DATABASE_URL")  # or other relevant config var
> if uri.startswith("postgres://"):
>     uri = uri.replace("postgres://", "postgresql://", 1)
> # rest of connection code using the connection string `uri`
> ```
> 
> https://help.heroku.com/ZKNTJQSK/why-is-sqlalchemy-1-4-x-not-connecting-to-heroku-postgres

Hello @cassiel257 - where do I insert this code? I tried putting it in my config.py but I keep getting the same error. I am deploying a flask app on heroku.

**Comment by cassiel257** [comment]
Hi @blacng - I inserted in my config file as below, and changed the name of a few variables to fit what was already in my project. I have the original DATABASE_URL defined in a separate file. Sometimes I have errors because of inadvertently adding spaces or single/double quotes. Can you share some details of the error, or your file setup?

![heroku postgres snip gh](https://user-images.githubusercontent.com/55888277/115131982-66182280-9fca-11eb-98f0-4e481fe0f37f.png)

**Comment by blacng** [comment]
> Hi @blacng - I inserted in my config file as below, and changed the name of a few variables to fit what was already in my project. I have the original DATABASE_URL defined in a separate file. Sometimes I have errors because of inadvertently adding spaces or single/double quotes. Can you share some details of the error, or your file setup?
> 
> ![heroku postgres snip gh](https://user-images.githubusercontent.com/55888277/115131982-66182280-9fca-11eb-98f0-4e481fe0f37f.png)

Thanks for your reply @cassiel257.

DATABASE_URL is defined in my docker-compose.yml

Below is my config.py:

![image](https://user-images.githubusercontent.com/13294689/115132813-a038f280-9fd1-11eb-8724-6cd7f197d452.png)

**Comment by cassiel257** [comment]
Hi @blacng,

I'm no YAML expert, so I don't know if this would help: have you tried
adding the same modification to the DevelopmentConfig, TestingConfig and
ProductionConfig classes? Are you receiving an error message in testing
config as well?

**Comment by blacng** [comment]
@cassiel257 - can you please show me a screenshot of the file where you defined your DATABASE_URL?

**Comment by cassiel257** [comment]
Hi @blacng, it was a relatively simple project, so locally I exported from a bash file, and on Heroku I manually entered the DATABASE_URL in the Config Vars section...the database path would of course be replaced by the heroku database instead of localhost...
![postgres bash snip](https://user-images.githubusercontent.com/55888277/115181360-c16b1300-a0a5-11eb-85c7-2a8a3b12abba.png)

**Comment by dchevell** [comment]
> can we "dupe" all these ? there were other issues opened w/ people surprised by the name removal right?

@zzzeek is there anywhere that captures the _why_ for deprecation? database URI's are something broader than SQLAlchemy and python, and I've run into situations in my org where we have orchestration layers injecting DB URI's into applications across many languages and frameworks - all of which continue to support the short form. Postgres' own docs still list it as valid.

I'm not here to argue the point (it's done now, and was deprecated long ago), but when asking internal teams to change to the long form at a layer that affects all languages/stacks/frameworks, the argument that one ORM decided to drop it doesn't hold water without some context as to why. I'm currently getting pushback from teams that consider SQLAlchemy "wrong" for dropping support for something that is still documented as supported by Postgres itself. I don't have a sufficient counter-argument, other than "switching to the long form would make my own life easier" 😛

**Comment by zzzeek** [comment]
> > can we "dupe" all these ? there were other issues opened w/ people surprised by the name removal right?
> 
> @zzzeek is there anywhere that captures the _why_ for deprecation? 

the reason the name was changed, which goes back to [here](https://docs.sqlalchemy.org/en/14/changelog/changelog_06.html#change-0ce9ee480a5a66cb9f77e0207f6dfc70), is because I had called the database dialect "postgres" because my original work with this database was in the 1995-ish kind of time, when that's what it was called.    Then around SQLAlchemy 0.5 or so, people told me hey, this database is actually called "PostgreSQL" now.   So I realized I had made a mistake, and i changed the name.

That's "why the change" but at that point the name "postgres" was legacy and we added a deprecation warning at that point.  "Why the deprecation" is because, it is normal to deprecate the old name of something when you fix the name to be something new.

> database URI's are something broader than SQLAlchemy and python,

SQLAlchemy URLs are not broader than SQLAlchemy.  They are specific to SQLAlchemy, do not look like anyone else's URLs (and they are more URLs than just URIs, I think) and they only work with SQLAlchemy.


> and I've run into situations in my org where we have orchestration layers injecting DB URI's into applications across many languages and frameworks - all of which continue to support the short form. 

I can't imagine using a single database URL without modification across many different application platforms.  I guess because PostgreSQL's libpq actually has a URL format that is fairly well known , there is some feasibility to this but certainly in the larger database world this is not a thing at all.      Like here is pymysql.connect: https://pymysql.readthedocs.io/en/latest/modules/connections.html  there is no URL there.   URLs for databases are not a universally available thing.


> Postgres' own docs still list it as valid.

I'm pretty sure this is up in the discussion here, that's the PostgreSQL database.  SQLAlchemy URLs are not libpq URIs.

> 
> I'm not here to argue the point (it's done now, and was deprecated long ago), but when asking internal teams to change to the long form at a layer that affects all languages/stacks/frameworks, the argument that one ORM decided to drop it doesn't hold water without some context as to why. 

I'm not following why that matters.    

> I'm currently getting pushback from teams that consider SQLAlchemy "wrong" for dropping support for something that is still documented as supported by Postgres itself. 

I don't understand the nature of this "pushback".   You didnt make this change, I did, and it's already done, so...what are they pushing against?    if these teams are in the business of using libpq URIs as a universal database connectivity format then shouldn't there be "pushback" that they aren't supporting MySQL, Oracle, SQL Server as well ?    If these teams are saying something like, "well we shouldnt support SQLAlchemy because they made this fickle decision"...that's too bad but honestly it makes them look a lot **more** fickle to make such a big deal over a change like this.

> I don't have a sufficient counter-argument, other than "switching to the long form would make my own life easier" stuck_out_tongue

what are....the two sides to this argument?  "sqlalchemy is wrong" / "sqlalchemy is right"  ...uhhh, water is wet / water is moist?  why does it matter?

**Comment by dchevell** [comment]
@zzzeek I'll apologise for any misunderstanding on my part - I guess the key point here is that I did not realise that the database URI is not intended to be a raw libpq URI. Given that it works, I had assumed there was an intention there, rather than it being a simple coincidence and therefore an SQLA-specific implementation detail. That's probably why the things I'm saying don't sound right to you. We've been injecting generic URI's into SQLA apps for a long time, and now this has suddenly broken I guess we didn't understand the intention here. Thanks for your insight Mike.

**Comment by zzzeek** [comment]
OK, I figured I'd do some research on this libpq URI issue.    Because first I noticed one thing, then I had kind of a weird suspicion, and then...the suspicion turned out to be ... true!

The feature was only added relatively recently in 9.2:  https://www.postgresql.org/docs/release/9.2.0/  which is two years after we deprecated "postgres".  it was also six years after SQLAlchemy's first release so our URL format predates theirs significantly.

Yet, as we see in the actual commit at https://github.com/postgres/postgres/commit/b035cb9db7aa7c0f28581b23feb10d3c559701f6#diff-66bf98128702c8fd0063e6989416d0d1739556d12f772d38a1845b83bd0acada , they seemed to focus on "postgres" but also added the comment (because there just wasnt enough irony already) "Oh, it also supports postgresql:// but that's probably just an accident." .     The URI format looks so much like ours that I crawled their dev list from that time to find where this was first proposed and sure enough our format was part of their consideration:

https://www.postgresql.org/message-id/4F45253C.4030409%402ndQuadrant.com

> First thing:  the URI prefix.  It is possible to connect using a URI in 
Python+SQL Alchemy, which was mentioned before as not too relevant due 
to their also requiring a driver name.  As documented at 
http://docs.sqlalchemy.org/en/latest/core/engines.html and demonstrated 
at http://packages.python.org/Flask-SQLAlchemy/config.html , it is 
possible to leave off the driver part of the connection string.  That 
assumes the default driver, such that postgresql:// does the same as 
postgresql+psycopg2:// , sensibly.  That means we absolutely have an 
installed base of URI speaking developers split between postgresql:// 
(Python) and postgres:// (Ruby).  Given that, there really isn't a 
useful path forward that helps out all those developers without 
supporting both prefixes.  That's where this left off before, I just 
wanted to emphasize how clear that need seems now.

So what we have above is exactly the rationale used by PG's developers to support both "postgresql://" as well as "postgres://".   **to be compatible with SQLAlchemy**.     So there you have it.    Their URL format **did not exist** when we deprecated "postgres://" and their use of "postgresql://" was at least somewhat based on their observations of us.   fin

**Comment by dchevell** [comment]
Hell of a find @zzzeek , and puts everything in a proper context. Thank you so much for digging about and clearing it up!

**Comment by slyapustin** [comment]
This issue is currently a blocker for Airflow auto-deployment on Heroku.
I created a workaround for that: https://github.com/apache/airflow/pull/24569

**Comment by zzzeek** [comment]
> This issue is currently a blocker for Airflow auto-deployment on Heroku. I created a workaround for that: [apache/airflow#24569](https://github.com/apache/airflow/pull/24569)

@slyapustin it's a blocker in...Airflow, you mean, I assume.    the definition of "blocker" is, "we can't do our work until project X fixes this problem, we have no workaround".   

if you want this issue to be pinged for third party libraries interacting with this change, just tag the SQLAlchemy issue link in your PR for the change on airflow, it will be logged.

**Comment by Dotrar** [comment]
Postgres allows for either: https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING-URIS

> The URI scheme designator can be either postgresql:// or postgres://. Each of the remaining URI parts is optional. 

as of version 16 of postgres. If the database specifies that both are fine, why not use both? or do a "find and replace" on your end rather than all of us having to do that same find and replace?

This update just broke my system, and now i have to talk to my platform provider and explain "yes yes i know postgres allows for both, but some person on the internet told me I have to only allow one and that happened to be not the one you chose" ???
